### PR TITLE
Implement `awaitOrTimeout`

### PR DESCRIPTION
**Pull request by mihails-strasuns-sociomantic**
Assuming I haven't missed any dangerous corner case, this should provide the base
of requested implementation.

### CODE REVIEW

File: src/ocean/task/util/Timer.d [hunk 1]
```diff
@@ -124,6 +124,113 @@ unittest
...
+    }
+
+    auto root = new RootTask;
+    root.to_wait_for = new InfiniteTask;
+
+    theScheduler.schedule(root);
+    theScheduler.eventLoop();
+
+    test(!root.used());
+    test(!root.to_wait_for.used());
```
Comment (nemanja-boric-sociomantic) [inline src/ocean/task/util/Timer.d]: Any way to write a quick test with waiting for the task that actually finishes?

Comment (nemanja-boric-sociomantic) [reply src/ocean/task/util/Timer.d]: (with no timeout)

Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/util/Timer.d]: Sure, will do (it would be a copy of one for `await` pretty much)

Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/util/Timer.d]: Good catch, I actually had one bug there!
Updated.

File: src/ocean/task/util/Timer.d [hunk 2]
```diff
@@ -124,6 +124,151 @@ unittest
...
+    task.terminationHook(&context.resume);
+    task.terminationHook(&scheduled_event.unregister);
+
+    // force async scheduling to avoid checking if this context needs
+    // suspend/resume and do it unconditionally
+    theScheduler.queue(task);
+    context.suspend();
+    // skip one scheduler event loop cycle to allow awaited task to finish with
+    // all termination hooks
+    theScheduler.processEvents();
```
Comment (nemanja-boric-sociomantic) [inline src/ocean/task/util/Timer.d]: Haha, yes of course :-)

Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/util/Timer.d]: Actually no, this is not good enough - the same fiber could be busy with other task by that time. I will update to provide required info by task itself.

**Comment by mihails-strasuns-sociomantic** [comment]
Updated with what should be more reliable approach to check if task is finished.

File: src/ocean/task/Scheduler.d [hunk 3]
```diff
@@ -685,6 +686,9 @@ final class Scheduler
         // calls `Task.continueAfterThrow()` and that throws again
         if (had_exception)
             this.processEvents();
+
+        // prevent referencing old worker fiber
+        task.fiber = null;
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/Scheduler.d:680]: Is this related to "Introduce task internal state bitmask"?

File: src/ocean/task/Task.d [hunk 4]
```diff
@@ -383,27 +384,42 @@ public abstract class Task : ISuspendable
...
     }
 
     /***************************************************************************
 
+        Returns:
+            true if this task was started and reached end of `run` method
+
+    ***************************************************************************/
+
+    final public bool finished ( )
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/Task.d:399]: Release note: new feature.

File: src/ocean/task/Task.d [hunk 5]
```diff
@@ -452,6 +471,7 @@ public abstract class Task : ISuspendable
         catch (TaskKillException)
         {
             debug_trace("<{}> termination (killed)", cast(void*) this);
+            this.state_bitmask &= ~TaskState.ToKill;
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/Task.d:474]: Interesting that `to_kill` was never reset anywhere before...!

File: relnotes/await-timeout.feature.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
+* `ocean.task.util.Timer`
+
+  New function `awaitOrTimeout` can be used to suspend some task until either
+  some other specified task finished OR until timer is hit, whichever comes
+  first. Conceptually is is similar to `theScheduler.await`, but with a timeout.
```
Comment (gavin-norman-sociomantic) [inline relnotes/await-timeout.feature.md]: Typo "it is"

File: relnotes/await-timeout.feature.md [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+  It is implemented as a free function in `Timer` module though to avoid adding
+  hard import dependency from scheduler module to all timer event code.
+
+  ```D
+  auto task = new SomeTask;
+  bool timeout = .awaitOrTimeout(task, 1_000_000);
+  // Doesn't kill on timeout by default, as it is not needed with tasks like DHT
```
Comment (gavin-norman-sociomantic) [inline relnotes/await-timeout.feature.md]: Though we're not prohibited from mentioning things like the DHT / swarm in public now, it feels weird to refer to it in ocean, which is a lower level library.

File: src/ocean/task/util/Timer.d [hunk 8]
```diff
@@ -124,6 +124,149 @@ unittest
 
 /*******************************************************************************
 
+    Similar to `theScheduler.await` but also has waiting timeout. Calling task
+    will be resumed either if awaited task finished or timeout is hit, whichever
+    happens first.
+
+    Params:
+        task = task to await for
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/util/Timer.d]: Grammar tip of the day: "await" or "wait for" (not "await for").

File: src/ocean/task/util/Timer.d [hunk 9]
```diff
@@ -124,6 +124,149 @@ unittest
...
+        return true;
+    }
+}
+
+///
+unittest
+{
+    initScheduler(SchedulerConfiguration.init);
+
+    .timer = new typeof(timer);
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/util/Timer.d]: Hmm seems unnecessary.

File: src/ocean/task/util/Timer.d [hunk 10]
```diff
@@ -83,7 +80,27 @@ unittest
 {
     initScheduler(SchedulerConfiguration.init);
 
-    .timer = new typeof(timer);
+    class SimpleTask : Task
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/util/Timer.d:83]: Is this related to "Create global timer lazily inside helper"?

**Review by gavin-norman-sociomantic — COMMENTED** [review]
Nice!

File: src/ocean/task/Scheduler.d (hunk 3 above)
Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/Scheduler.d:680]: Yes, though a bit indirectly. `task.fiber is null` with thi can be used to check for extra state - "not used at all".

File: src/ocean/task/Task.d (hunk 5 above)
Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/Task.d:474]: Don't bring attention to my failures gone unnoticed :P

File: src/ocean/task/util/Timer.d (hunk 10 above)
Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/util/Timer.d:83]: Yes, diff is confusing. I have split unittest that requires resetting timer instance manually to check allocated slot count from documented unittest which uses regular lazily created timer.

**Comment by mihails-strasuns-sociomantic** [comment]
Updated with doc and release note fixes

**Comment by mihails-strasuns-sociomantic** [comment]
Ping @gavin-norman-sociomantic

**Comment by mihails-strasuns-sociomantic** [comment]
Ping @gavin-norman-sociomantic

File: src/ocean/task/util/Timer.d (hunk 9 above)
Comment (gavin-norman-sociomantic) [reply src/ocean/task/util/Timer.d]: Unnecessary?

**Review by gavin-norman-sociomantic — COMMENTED** [review]
Just the question about initialising the timer in unittests.

File: src/ocean/task/util/Timer.d (hunk 9 above)
Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/util/Timer.d]: Necessay until timer setup code gets rewritten in next commit. Depending on order of test block being run, `.timer` could be null here.

File: src/ocean/task/util/Timer.d [hunk 11]
```diff
@@ -83,7 +80,27 @@ unittest
...
+    theScheduler.eventLoop();
+}
+
+unittest
+{
+    // same as previous, but test allocated slot count
+
+    initScheduler(SchedulerConfiguration.init);
+
+    .timer = new typeof(.timer);
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/util/Timer.d]: And is it necessary here?

Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/util/Timer.d]: Yes, for a reason explained above (it tests exact allocated slot count)

Comment (gavin-norman-sociomantic) [reply src/ocean/task/util/Timer.d]: Hmm, I don't understand that comment. What's a "slot"? Why is it necessary to re-allocate the timer? Doc enhancement please :)

Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/util/Timer.d]: https://github.com/sociomantic-tsunami/ocean/pull/234/files#diff-a3ce7f2ae79a2097ba597486e0d9a30fR121 + comment 3 lines above

Which doc?

Comment (gavin-norman-sociomantic) [reply src/ocean/task/util/Timer.d]: Some documentation of what this unittest is testing. It's not clear to me.

>Why is it necessary to re-allocate the timer?

"Reallocate a fresh timer so the test result is not affected by previous unittests" ?

Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/util/Timer.d]: Updated

**Review by gavin-norman-sociomantic — APPROVED** [review]

**Pull request merged**